eliot/_action.py

```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
+
+    @ivar TaskLevel task_level: The level of the task in which the action occurs.
+
+    @ivar start_time: The Unix timestamp of when the action started.
+
+    @ivar end_time: The Unix timestamp of when the action ended, or C{None} if
+        there has been no end message.
+
+    @ivar exception: If the action failed, the name of the exception that was
+        raised to cause it to fail. If the action succeeded, or hasn't finished
+        yet, then C{None}.
+
+    @ivar reason: The reason the action failed. If the action succeeded, or
+        hasn't finished yet, then C{None}.
+
+    @ivar _children: A L{pvector} of L{WrittenAction} and L{WrittenMessage}
+        that make up this action.
+    """
+
+    # Broad philosophy:
+    # - Have a way of constructing known-valid instances (at the very least,
+    #   this is useful for tests)
+    # - Have a thing that takes messages out of order and hangs on to them
+    #   until it can construct known-valid instances
+    # - Be able to recover the original messages
+
+    action_type = field(type=unicode, mandatory=True)
+    status = field(type=unicode, mandatory=True,
+                   invariant=lambda x: (x in VALID_STATUSES, 'Invalid status'))
+    task_uuid = field(type=UUID, mandatory=True)
+    task_level = field(type=TaskLevel, mandatory=True)
+    start_time = field(type=float, mandatory=True)
+    end_time = field(type=optional(float), mandatory=True)
+    # XXX: Actually a map to either WrittenMessage or WrittenAction, but
+    # pyrsistent doesn't support recursive data types:
+    # https://github.com/tobgu/pyrsistent/issues/48
+    _children = pmap_field(TaskLevel, object)
+    exception = field(type=optional(unicode), mandatory=True)
+    reason = field(type=optional(unicode), mandatory=True)
+
+    # XXX: Possible invariants:
+    # - start_time <= end_time
+    # - task_level = set(child.task_level.parent() for child in children)
+    # - task_uuid = set(child.task_uuid for child in children)
+    # - exception and reason set iff status == FAILED_STATUS
+
+    # XXX: I'm fairly convinced that this would be simpler and clearer if we
+    # had a class or classes for the "special" action messages. Also if we had
+    # a way of expressing 'either a WrittenMessage or a WrittenAction' more
+    # succinctly.
+
+    @classmethod
+    def from_messages(cls, start_message, children=v(), end_message=None):
+        """
+        Create a C{WrittenAction} from C{WrittenMessage}s and other
+        C{WrittenAction}s.
+
+        @param WrittenMessage start_message: A message that has
+            C{ACTION_STATUS_FIELD}, C{ACTION_TYPE_FIELD}, and a C{task_level}
+            that ends in C{1}.
+        @param children: An iterable of C{WrittenMessage} and C{WrittenAction}
+        @param WrittenMessage end_message: A message that has the same
+            C{action_type} as this action.
+
+        @raise WrongTask: If C{end_message} has a C{task_uuid} that differs
```

Comment:
I wonder if having individual exceptions is sufficiently useful that it's not worth just simplifying things by relying on pyrsistent invariants.